src/Raven.Server/Documents/Attachments/AttachmentUploader.cs

```diff
@@ -0,0 +1,191 @@
+                }
+
+                attachmentStream.CopyTo(stream);
+
+                TaskCancelToken.Token.ThrowIfCancellationRequested();
+            }
+        },
+        TaskCancelToken.Token,
+        taskOptions);
+
+        return task;
+    }
+
+    public async Task<bool> WaitForFinishedTasksIfNeededAsync(Stopwatch sp, OperationCancelToken token)
+    {
+        if (_threads.Count < _concurrentThreads)
+            return true;
+
+        while (_threads.Count >= _concurrentThreads)
+        {
+            if (token.Token.IsCancellationRequested)
+                return false;
+
+            if (sp.ElapsedMilliseconds > RetireAttachmentsSender.ReadTransactionMaxOpenTimeInMs)
+                return false;
+
+            // Create a timeout task for the polling interval
+            await Task.Delay(512);
+
+            var current = _threads.First;
+            while (current != null)
+            {
+                if (token.Token.IsCancellationRequested)
+                    return false;
+
+                var next = current.Next; // Store next before potential removal
+
+                if (AssertTaskStateAndInvokeAction(current.Value))
+                {
+                    _threads.Remove(current); // O(1) operation for LinkedListNode
+                }
+
+                current = next;
+            }
+        }
+
+        if (token.Token.IsCancellationRequested)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    private bool AssertTaskStateAndInvokeAction(AttachmentUploadToCloudHolder task)
+    {
+        if (task.UploadTask.IsCompleted)
+        {
+            if (task.UploadTask.IsCompletedSuccessfully)
+            {
+                OnSuccessAction?.Invoke(task);
+            }
+            else if (task.UploadTask.IsFaulted)
+            {
+                if (_logger.IsErrorEnabled)
+                {
+                    _logger.Error($"Upload task of retired attachment '{task.Doc.LowerId}' with identifier '{task.Doc.Id}' failed.", task.UploadTask.Exception);
+                }
+
+                OnExceptionAction?.Invoke(task);
+            }
+            else if (task.UploadTask.IsCanceled)
+            {
+                if (_logger.IsDebugEnabled)
+                {
+                    _logger.Debug($"Upload task of retired attachment '{task.Doc.LowerId}' with identifier '{task.Doc.Id}' was canceled.");
+                }
+            }
+
+            return true;
+        }
+
+        return false;
+    }
+
+    public override void Execute()
+    {
+        foreach (var t in _threads)
+        {
+            TaskCancelToken.Token.ThrowIfCancellationRequested();
+
+            AsyncHelpers.RunSync(() => t.UploadTask);
+            AssertTaskStateAndInvokeAction(t);
+        }
+
+        _threads.Clear();
+    }
+
+    private static long GetAttachmentSizeThreshold()
+    {
+        if (PlatformDetails.Is32Bits)
+            return 64 * 1024 * 1024;
```

Comment:
use Sparrow.Global.Constants.Size

Reply:
also for 32 bits its needs to be way less, probably 8mb max. maybe 16

Reply:
done